core/emulate.c

```diff
@@ -0,0 +1,978 @@
+        .decode_dst   = &decode_##_dec_dst,   \
+        .decode_src1  = &decode_##_dec_src1,  \
+        .decode_src2  = &decode_##_dec_src2,  \
+        .flags        = _flags | INSN_GROUP | INSN_MODRM \
+    }
+
+#define \
+    F(_handler, _dec_dst, _dec_src1, _dec_src2, _flags) \
+    I(_handler, _dec_dst, _dec_src1, _dec_src2, _flags | INSN_FASTOP)
+
+#define I2_BV(_handler, _dec_dst, _dec_src1, _dec_src2, _flags) \
+    I(_handler, _dec_dst, _dec_src1, _dec_src2, (_flags | INSN_BYTEOP)), \
+    I(_handler, _dec_dst, _dec_src1, _dec_src2, (_flags))
+#define F2_BV(_handler, _dec_dst, _dec_src1, _dec_src2, _flags) \
+    F(_handler, _dec_dst, _dec_src1, _dec_src2, (_flags | INSN_BYTEOP)), \
+    F(_handler, _dec_dst, _dec_src1, _dec_src2, (_flags))
+
+#define F6_ALU(_handler, _flags) \
+    F2_BV(_handler, op_modrm_rm, op_modrm_reg, op_none, (_flags | INSN_MODRM)), \
+    F2_BV(_handler, op_modrm_reg, op_modrm_rm, op_none, (_flags | INSN_MODRM)), \
+    F2_BV(_handler, op_acc, op_imm, op_none, (_flags))
+
+/* Soft-emulation */
+static void em_mov(struct em_context_t *ctxt);
+static void em_movzx(struct em_context_t *ctxt);
+static void em_xchg(struct em_context_t *ctxt);
+
+static const struct em_opcode_t opcode_group1[8] = {
+    F(em_add, op_none, op_none, op_none, 0),
+    F(em_or,  op_none, op_none, op_none, 0),
+    F(em_adc, op_none, op_none, op_none, 0),
+    F(em_sbb, op_none, op_none, op_none, 0),
+    F(em_and, op_none, op_none, op_none, 0),
+    F(em_sub, op_none, op_none, op_none, 0),
+    F(em_xor, op_none, op_none, op_none, 0),
+    F(em_cmp, op_none, op_none, op_none, 0),
+};
+
+static const struct em_opcode_t opcode_group3[8] = {
+    F(em_test, op_modrm_rm, op_imm, op_none, 0),
+    F(em_test, op_modrm_rm, op_imm, op_none, 0),
+    F(em_not, op_modrm_rm, op_none, op_none, 0),
+    F(em_neg, op_modrm_rm, op_none, op_none, 0),
+};
+
+static const struct em_opcode_t opcode_group11[8] = {
+    I(em_mov, op_none, op_none, op_none, INSN_MOV),
+};
+
+static const struct em_opcode_t opcode_table[256] = {
+    /* 0x00 - 0x07 */
+    F6_ALU(em_add, 0), X2(N),
+    /* 0x08 - 0x0F */
+    F6_ALU(em_or, 0),  X2(N),
+    /* 0x10 - 0x17 */
+    F6_ALU(em_adc, 0), X2(N),
+    /* 0x18 - 0x1F */
+    F6_ALU(em_sbb, 0), X2(N),
+    /* 0x20 - 0x27 */
+    F6_ALU(em_and, 0), X2(N),
+    /* 0x28 - 0x2F */
+    F6_ALU(em_sub, 0), X2(N),
+    /* 0x30 - 0x37 */
+    F6_ALU(em_xor, 0), X2(N),
+    /* 0x38 - 0x3F */
+    F6_ALU(em_cmp, 0), X2(N),
+    /* 0x40 - 0x47 */
+    X8(F(em_inc, op_modrm_reg, op_none, op_none, 0)),
+    /* 0x48 - 0x4F */
+    X8(F(em_dec, op_modrm_reg, op_none, op_none, 0)),
+    /* 0x50 - 0x7F */
+    X16(N), X16(N), X16(N),
+    /* 0x80 - 0x8F */
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, 0),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_simm8, op_none, 0),
```

Comment:
I checked all 8 instructions in this group. 0x80-0x83 all require the immediate operand to be sign-extended to the operand size. So instead of `op_simm8`, we probably need `op_simm`?